### ISSUE TITLE
[Bug] Emails scheduled by campaign are deleted after second lead triggers the same event

### ISSUE THREAD

**Issue by alvinbenavides** [mautic/mautic#150]
I don’t know how to start explaining this one but I’ll try my best…

I have a campaign where after downloading an asset, two emails are scheduled to be sent to the lead. The first after 10 minutes and the second after 20 minutes.

Right now, when a lead downloads the asset, emails are scheduled and sent correctly on the specified time... Triggers and events are working fine.

The problem comes when a second lead downloads the asset before the scheduled emails for the first lead are sent. When this happen, emails are scheduled for the second lead but are also removed for the first one.

Basically:
1) Lead 1 downloads asset
2) Campaign schedule two emails for Lead 1
3) Lead 2 downloads asset before emails for Lead 1 are sent
4) Campaign schedule two emails for Lead 2
5) Scheduled emails for Lead 1 disappear from schedule.

First I notice this testing from the same IP and then test it from two different devices with different public IP but still the same.

If you need more explanation or info please let me know. Thanks

**Comment by alanhartless** [comment]
Hi Alvin,

First, apply the changes in https://github.com/mautic/mautic/pull/154.

Delete app/cache/prod

Then open index.php and temporarily change 

```
$kernel = new AppKernel('prod', false);
```
to 
```
$kernel = new AppKernel('prod', true);
```

That puts the kernel in debug mode and one of the changes sets logging into debug mode as well.

Then, try to replicate the scenario above.  It should write a bunch of campaign execution stuff to the log.  Can you post those?

I tried to replicate the issue and it worked fine for me.  One thing I'm curious about, for lead 1 and lead 2, can you check your lead_ips_xref table and see if one of the leads are associated with both public IPs you used?

Are you able to view the cookies of the both devices you used?  If so, can you browse to your mautic site and view the cookies.  Do you have more than one cookies one either device  that is a random string but the value is a the ID of your leads? 

Thanks,
Alan

**Comment by alvinbenavides** [comment]
Hi Alan. 

Since I'm running a test environment, I'm actually creating leads with my emails accounts and forwarders. Then deleting them in order to create "new" leads. I always delete cookies between leads.

Also, keeping in mind the error about point triggers we've being discussing, after making the changes you mentioned and based on your questions, I made 3 different scenarios.

I stared the test deleting points actions (in the original scenario opening emails add points to leads). I also test it with two email addresses that I've never used.

1) No point actions, new email addresses. This time scheduled emails where not deleted and everything works as expected.

2) No points actions, previous "leads" deleted from DB and used again. This time scheduled emails where not deleted and everything works as expected.

3) Added a Point Action that adds one point to the lead after opening an email. This time the error was repeated, scheduled emails from Lead 1 were deleted.

I'm sending you the logs by email. If you need any other info please let me know.

Thanks

**Comment by alvinbenavides** [comment]
Hi Alan. 

Since I'm running a test environment, I'm actually creating leads with
my emails accounts and forwarders. Then deleting them in order to create
"new" leads. I always delete cookies between leads. 

Also, keeping in mind the error about point triggers we've being
discussing, after making the changes you mentioned and based on your
questions, I made 3 different scenarios. 

I stared the test deleting points actions (in the original scenario
opening emails add points to leads). I also test it with two email
addresses that I've never used. 

1) No point actions, new email addresses. This time scheduled emails
where not deleted and everything works as expected. 

2) No points actions, previous "leads" deleted from DB and used again.
This time scheduled emails where not deleted and everything works as
expected. 

3) Added a Point Action that adds one point to the lead after opening an
email. This time the error was repeated, scheduled emails from Lead 1
were deleted. 

I'm sending you the logs by email. If you need any other info please let
me know. 

Thanks 

El 2015-03-07 22:46, Alan Hartless escribió: 

> Hi Alvin, 
> 
> First, apply the changes in #154 [1]. 
> 
> Delete app/cache/prod 
> 
> Then open index.php and temporarily change 
> 
> $kernel = new AppKernel('prod', false);
> 
> to 
> 
> $kernel = new AppKernel('prod', true);
> 
> That puts the kernel in debug mode and one of the changes sets logging into debug mode as well. 
> 
> Then, try to replicate the scenario above. It should write a bunch of campaign execution stuff to the log. Can you post those? 
> 
> I tried to replicate the issue and it worked fine for me. One thing I'm curious about, for lead 1 and lead 2, can you check your lead_ips_xref table and see if one of the leads are associated with both public IPs you used? 
> 
> Are you able to view the cookies of the both devices you used? If so, can you browse to your mautic site and view the cookies. Do you have more than one cookies one either device that is a random string but the value is a the ID of your leads? 
> 
> Thanks,
> Alan 
> 
> --
> Reply to this email directly or view it on GitHub [2].

 

Links:
------
[1] https://github.com/mautic/mautic/pull/154
[2] https://github.com/mautic/mautic/issues/150#issuecomment-77733047

**Comment by alanhartless** [comment]
This is likely all related to 178 which fixes the lead list issue we've been working through.  What is happening is that the user is added to the campaign through your lead lists.  When the lead is removed from the list, they are also removed from the campaign thus deleting any scheduled emails. 

That PR should finally fix leads from getting incorrectly removed from the list (point changes triggered the leads lists to be reexamined).  So, it should also leave the campaign and the scheduled emails alone.

**Comment by PatchRanger** [comment]
It's an old issue - but I guess I have a regression. I am persistently able to reproduce the bug on 1.4.1. Right now I've turned aforementioned debug mode on in order to collect some valuable data to debug this behavior.
I have a campaign with "Send email" and "Change tags" actions together with "Opens email" decision. It has about 30 steps once in a day.
The symptoms are absolutely the same: when contacts enter the campaign, they got scheduled events, everything is ok. But I've noticed that there is no persistent flow from early steps of campaign to further ones: I've got 300 contacts in 1st step, 5 in 2nd, 3 in 3rd and no in further (though there should be) - which definitely looks weird, right?
I am going to write follow-up once I get insight by looking into campaign logs.

**Comment by PatchRanger** [comment]
Very much similar issue: #2004.

**Comment by PatchRanger** [comment]
I didn't get much from the logs.
But I've noticed that after changing filter conditions for segment its contact count gets decreased a lot - and then replenished. I guess, the bug caused by the rebuilding of campaign leads: they got removed first (even if they match new condition) and then added again (which proves that they obey to filter and shouldn't be removed), loosing scheduled events.

**Comment by edutj** [comment]
Any solutions for it?

**Comment by PatchRanger** [comment]
For now I've just commented `removeScheduledEvents` call in `CampaignModel::removeLeads`. Pros: my campaigns stopped to loose scheduled events. Cons: I guess, my hack would leave scheduled events untouched (wouldn't remove) in case if a contact leaves campaign. For me it's a reasonable price at the moment as I don't know how to fix it "natural way".

**Comment by edutj** [comment]
Yeah. I don't think this hack would solve my issue #2004 .

**Comment by PatchRanger** [comment]
@edutj It should, give it a try) It works in my case. The line I've commented is the only place in code, which deletes scheduled events - so the bug for sure is just call to appropriate function in inappropriate moment. While we're figuring out why it is so, my choice is just to skip deleting of scheduled events. Not a solution but still a resolution)

**Comment by edutj** [comment]
@PatchRanger do you think if I comment removeScheduledEvents call in CampaignModel::removeLeads as you said, the already removed ones will come back?

**Comment by PatchRanger** [comment]
@edutj No, I am afraid, it looks like resurrection of already removed events is not possible (I could be wrong). You could clone your campaign and put all of these "unfinished" contacts into it - but it means that the campaign for them would be restarted from scratch, so it fits only you feel ok about restarting.
It is here: https://github.com/mautic/mautic/blob/staging/app/bundles/CampaignBundle/Model/CampaignModel.php#L815.

**Comment by edutj** [comment]
Yeah. I noticed that IF i clone the campaign, it works. But I think I'll move part of my email operations to another service. Specially because I can't manually move a lead inside the campaign (if Mautic allowed that, i could clone the campaign and just move the users around inside the campaign).
